icaruscode/Utilities/detectorclocks_icarus.fcl

```diff
@@ -25,15 +25,15 @@ icarus_detectorclocks: {
 
     TrigModuleName:    ""
     InheritClockConfig: true
-    G4RefTime:        -1.15e3  # G4 time [us] where electronics clock counting start
+    G4RefTime:        -1.5e3  # G4 time [us] where electronics clock counting start
```

Comment:
This looks like a global change that will break all existing simulation. Why do you think we need it?

Changing `DefaultTrigTime` changes the value of `TriggerTime()` from `1150_us` to `1500_us` from the electronics time start. A particle generated at simulation time `0` produces activity (e.g. PMT waveform timestamp) at electronics time 1150 µs. When we match it to generation, we take that timestamp and convert it into generator time: remove `TriggerTime()`. If we do that in the middle, we'll try to match PMT activity at 1150 µs electronics time to generated particles at 1150 - 1500 = -350 µs.

Reply:
This change was intended to account for the change to the corsika generation window. I thought that since I had moved the start of the cosmic window to -1.5ms (relative to the TPC start), I also needed to move the G4RefTime back to -1.5ms to make sure that the cosmics were simulated correctly. I also believed that I had accounted for all the timings so that this change wouldn't break other things, but I definitely could have missed something.

Are either of these thoughts wrong?

Reply:
I am not 100% sure, but I think so. `G4RefTime` is a relative time offset between the `0` of the generation time and the `0` of the electronics time scale (check [`DetectorClocks` documentation](https://nusoft.fnal.gov/larsoft/doxsvn/html/classdetinfo_1_1DetectorClocks.html#details) if you are looking for an headache). What you have called "TPC start" is in there called _beam gate opening_ (or _hardware trigger time_, because... because of confusion, probably).
You did not change neither the electronics time scale reference (which I argue you should not change to preserve existing data) nor the generation time scale reference, since its `0` is still in the same place (when the beam arrives).
The only difference with the new generation is that there will be activity at a time that is measured by negative electronics time scale numbers. Is that a problem? well, _it should not_, and furthermore this should not get anywhere beyond digitisation, since the readout time range is well within the positive side. Of course, "it should not" does not answer the question "is it?".
In terms of changing everything not to break things: you have changed, as far as I can tell, all it was needed to have a consistent simulation. But the change of reference time makes the new, consistent configuration incompatible with the data generated with the older, also consistent, configuration.

Reply:
Gotcha -- I will check to see if anything goes wrong if I leave this timing config unchanged.

Reply:
I have investigated this further and I believe that the timing changes __are__ needed in order to accommodate the changes to the corsika window. The reason is that existing larsoft code seems to make the assumption that no particles are generated with an electronics time <0. 

I followed up with the document Tracy commented about below -- it contains the line: 

> In simulated events, the electronics clock is related to geant time by a fixed offset.
The start time of the simulated electronics clock is required to precede any truth time
in the event, but is otherwise arbitrary.

I followed up on this quote with Herbert Greenlee, and he confirmed that he didn't think a configuration where the G4RefTime is later than the particle generation time would work. He believed that it would cause issues with G4. 

I have not been able to find any specific problems in G4, but I have been able to find pieces of code that assume a non-negative electronics time. For example, this snippet of code in `larsim/LegacyLArG4/LArVoxelReadout.cxx`:

            /// \todo check on what happens if we allow the tdc value to be
            /// \todo beyond the end of the expected number of ticks
            // Add potential decay/capture/etc delay effect, simTime.
            unsigned int tdc = fClock.Ticks(ts->G4ToElecTime(TDiff + simTime));
            // Add electrons produced by each cluster to the map
            DepositsToStore[channel][tdc].add(nEnDiff[k], nElDiff[k]);

If there is a particle generated at negative electronics time with charge near the wires, then the `tdc` variable would underflow. And, as mentioned by the comment, it does not seem to be tested behavior for `tdc` to exceed the expected readout window.

I've tested out running gen+G4 with the new corsika window and the old timing settings, and the code did not crash. However, from looking through the code and prior documentation, it seems like code does make the assumption that there are no particles at negative electronics time. So I would be worried about any logic errors in the output.

Anyway, let me know what you think! And if there are any questions / anything I should follow up on.